// botframework-webchat-api/lib/hooks/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "useActiveTyping", {
  enumerable: true,
  get: function get() {
    return _useActiveTyping.default;
  }
});
Object.defineProperty(exports, "useActivities", {
  enumerable: true,
  get: function get() {
    return _useActivities.default;
  }
});
Object.defineProperty(exports, "useActivityKeys", {
  enumerable: true,
  get: function get() {
    return _useActivityKeys.default;
  }
});
Object.defineProperty(exports, "useActivityKeysByRead", {
  enumerable: true,
  get: function get() {
    return _useActivityKeysByRead.default;
  }
});
Object.defineProperty(exports, "useAvatarForBot", {
  enumerable: true,
  get: function get() {
    return _useAvatarForBot.default;
  }
});
Object.defineProperty(exports, "useAvatarForUser", {
  enumerable: true,
  get: function get() {
    return _useAvatarForUser.default;
  }
});
Object.defineProperty(exports, "useByteFormatter", {
  enumerable: true,
  get: function get() {
    return _useByteFormatter.default;
  }
});
Object.defineProperty(exports, "useConnectivityStatus", {
  enumerable: true,
  get: function get() {
    return _useConnectivityStatus.default;
  }
});
Object.defineProperty(exports, "useCreateActivityRenderer", {
  enumerable: true,
  get: function get() {
    return _useCreateActivityRenderer.default;
  }
});
Object.defineProperty(exports, "useCreateActivityStatusRenderer", {
  enumerable: true,
  get: function get() {
    return _useCreateActivityStatusRenderer.default;
  }
});
Object.defineProperty(exports, "useCreateAttachmentForScreenReaderRenderer", {
  enumerable: true,
  get: function get() {
    return _useCreateAttachmentForScreenReaderRenderer.default;
  }
});
Object.defineProperty(exports, "useCreateAvatarRenderer", {
  enumerable: true,
  get: function get() {
    return _useCreateAvatarRenderer.default;
  }
});
Object.defineProperty(exports, "useCreateScrollToEndButtonRenderer", {
  enumerable: true,
  get: function get() {
    return _useCreateScrollToEndButtonRenderer.default;
  }
});
Object.defineProperty(exports, "useDateFormatter", {
  enumerable: true,
  get: function get() {
    return _useDateFormatter.default;
  }
});
Object.defineProperty(exports, "useDebouncedNotifications", {
  enumerable: true,
  get: function get() {
    return _useDebouncedNotifications.default;
  }
});
Object.defineProperty(exports, "useDictateInterims", {
  enumerable: true,
  get: function get() {
    return _useDictateInterims.default;
  }
});
Object.defineProperty(exports, "useDictateState", {
  enumerable: true,
  get: function get() {
    return _useDictateState.default;
  }
});
Object.defineProperty(exports, "useDirection", {
  enumerable: true,
  get: function get() {
    return _useDirection.default;
  }
});
Object.defineProperty(exports, "useDisabled", {
  enumerable: true,
  get: function get() {
    return _useDisabled.default;
  }
});
Object.defineProperty(exports, "useDismissNotification", {
  enumerable: true,
  get: function get() {
    return _useDismissNotification.default;
  }
});
Object.defineProperty(exports, "useEmitTypingIndicator", {
  enumerable: true,
  get: function get() {
    return _useEmitTypingIndicator.default;
  }
});
Object.defineProperty(exports, "useGetActivityByKey", {
  enumerable: true,
  get: function get() {
    return _useGetActivityByKey.default;
  }
});
Object.defineProperty(exports, "useGetHasAcknowledgedByActivityKey", {
  enumerable: true,
  get: function get() {
    return _useGetHasAcknowledgedByActivityKey.default;
  }
});
Object.defineProperty(exports, "useGetKeyByActivity", {
  enumerable: true,
  get: function get() {
    return _useGetKeyByActivity.default;
  }
});
Object.defineProperty(exports, "useGetKeyByActivityId", {
  enumerable: true,
  get: function get() {
    return _useGetKeyByActivityId.default;
  }
});
Object.defineProperty(exports, "useGetSendTimeoutForActivity", {
  enumerable: true,
  get: function get() {
    return _useGetSendTimeoutForActivity.default;
  }
});
Object.defineProperty(exports, "useGrammars", {
  enumerable: true,
  get: function get() {
    return _useGrammars.default;
  }
});
Object.defineProperty(exports, "useGroupActivities", {
  enumerable: true,
  get: function get() {
    return _useGroupActivities.default;
  }
});
Object.defineProperty(exports, "useGroupTimestamp", {
  enumerable: true,
  get: function get() {
    return _useGroupTimestamp.default;
  }
});
Object.defineProperty(exports, "useLanguage", {
  enumerable: true,
  get: function get() {
    return _useLanguage.default;
  }
});
Object.defineProperty(exports, "useLastAcknowledgedActivityKey", {
  enumerable: true,
  get: function get() {
    return _useLastAcknowledgedActivityKey.default;
  }
});
Object.defineProperty(exports, "useLastReadActivityKey", {
  enumerable: true,
  get: function get() {
    return _useLastReadActivityKey.default;
  }
});
Object.defineProperty(exports, "useLastTypingAt", {
  enumerable: true,
  get: function get() {
    return _useLastTypingAt.default;
  }
});
Object.defineProperty(exports, "useLocalize", {
  enumerable: true,
  get: function get() {
    return _useLocalize.default;
  }
});
Object.defineProperty(exports, "useLocalizeDate", {
  enumerable: true,
  get: function get() {
    return _useLocalizeDate.default;
  }
});
Object.defineProperty(exports, "useLocalizer", {
  enumerable: true,
  get: function get() {
    return _useLocalizer.default;
  }
});
Object.defineProperty(exports, "useMarkActivityAsSpoken", {
  enumerable: true,
  get: function get() {
    return _useMarkActivityAsSpoken.default;
  }
});
Object.defineProperty(exports, "useMarkActivityKeyAsRead", {
  enumerable: true,
  get: function get() {
    return _useMarkActivityKeyAsRead.default;
  }
});
Object.defineProperty(exports, "useMarkAllAsAcknowledged", {
  enumerable: true,
  get: function get() {
    return _useMarkAllAsAcknowledged.default;
  }
});
Object.defineProperty(exports, "usePerformCardAction", {
  enumerable: true,
  get: function get() {
    return _usePerformCardAction.default;
  }
});
Object.defineProperty(exports, "usePostActivity", {
  enumerable: true,
  get: function get() {
    return _usePostActivity.default;
  }
});
Object.defineProperty(exports, "useReferenceGrammarID", {
  enumerable: true,
  get: function get() {
    return _useReferenceGrammarID.default;
  }
});
Object.defineProperty(exports, "useRelativeTimeFormatter", {
  enumerable: true,
  get: function get() {
    return _useRelativeTimeFormatter.default;
  }
});
Object.defineProperty(exports, "useRenderActivity", {
  enumerable: true,
  get: function get() {
    return _useRenderActivity.default;
  }
});
Object.defineProperty(exports, "useRenderActivityStatus", {
  enumerable: true,
  get: function get() {
    return _useRenderActivityStatus.default;
  }
});
Object.defineProperty(exports, "useRenderAttachment", {
  enumerable: true,
  get: function get() {
    return _useRenderAttachment.default;
  }
});
Object.defineProperty(exports, "useRenderAvatar", {
  enumerable: true,
  get: function get() {
    return _useRenderAvatar.default;
  }
});
Object.defineProperty(exports, "useRenderToast", {
  enumerable: true,
  get: function get() {
    return _useRenderToast.default;
  }
});
Object.defineProperty(exports, "useRenderTypingIndicator", {
  enumerable: true,
  get: function get() {
    return _useRenderTypingIndicator.default;
  }
});
Object.defineProperty(exports, "useSendBoxValue", {
  enumerable: true,
  get: function get() {
    return _useSendBoxValue.default;
  }
});
Object.defineProperty(exports, "useSendEvent", {
  enumerable: true,
  get: function get() {
    return _useSendEvent.default;
  }
});
Object.defineProperty(exports, "useSendFiles", {
  enumerable: true,
  get: function get() {
    return _useSendFiles.default;
  }
});
Object.defineProperty(exports, "useSendMessage", {
  enumerable: true,
  get: function get() {
    return _useSendMessage.default;
  }
});
Object.defineProperty(exports, "useSendMessageBack", {
  enumerable: true,
  get: function get() {
    return _useSendMessageBack.default;
  }
});
Object.defineProperty(exports, "useSendPostBack", {
  enumerable: true,
  get: function get() {
    return _useSendPostBack.default;
  }
});
Object.defineProperty(exports, "useSendTimeoutForActivity", {
  enumerable: true,
  get: function get() {
    return _useSendTimeoutForActivity.default;
  }
});
Object.defineProperty(exports, "useSendTypingIndicator", {
  enumerable: true,
  get: function get() {
    return _useSendTypingIndicator.default;
  }
});
Object.defineProperty(exports, "useSetNotification", {
  enumerable: true,
  get: function get() {
    return _useSetNotification.default;
  }
});
Object.defineProperty(exports, "useShouldSpeakIncomingActivity", {
  enumerable: true,
  get: function get() {
    return _useShouldSpeakIncomingActivity.default;
  }
});
Object.defineProperty(exports, "useStartDictate", {
  enumerable: true,
  get: function get() {
    return _useStartDictate.default;
  }
});
Object.defineProperty(exports, "useStopDictate", {
  enumerable: true,
  get: function get() {
    return _useStopDictate.default;
  }
});
Object.defineProperty(exports, "useStyleOptions", {
  enumerable: true,
  get: function get() {
    return _useStyleOptions.default;
  }
});
Object.defineProperty(exports, "useSubmitSendBox", {
  enumerable: true,
  get: function get() {
    return _useSubmitSendBox.default;
  }
});
Object.defineProperty(exports, "useSuggestedActions", {
  enumerable: true,
  get: function get() {
    return _useSuggestedActions.default;
  }
});
Object.defineProperty(exports, "useTimeoutForSend", {
  enumerable: true,
  get: function get() {
    return _useTimeoutForSend.default;
  }
});
Object.defineProperty(exports, "useTrackDimension", {
  enumerable: true,
  get: function get() {
    return _useTrackDimension.default;
  }
});
Object.defineProperty(exports, "useTrackEvent", {
  enumerable: true,
  get: function get() {
    return _useTrackEvent.default;
  }
});
Object.defineProperty(exports, "useTrackException", {
  enumerable: true,
  get: function get() {
    return _useTrackException.default;
  }
});
Object.defineProperty(exports, "useTrackTiming", {
  enumerable: true,
  get: function get() {
    return _useTrackTiming.default;
  }
});
Object.defineProperty(exports, "useUserID", {
  enumerable: true,
  get: function get() {
    return _useUserID.default;
  }
});
Object.defineProperty(exports, "useUsername", {
  enumerable: true,
  get: function get() {
    return _useUsername.default;
  }
});
Object.defineProperty(exports, "useVoiceSelector", {
  enumerable: true,
  get: function get() {
    return _useVoiceSelector.default;
  }
});

var _useActiveTyping = _interopRequireDefault(require("./useActiveTyping"));

var _useActivities = _interopRequireDefault(require("./useActivities"));

var _useActivityKeys = _interopRequireDefault(require("./useActivityKeys"));

var _useActivityKeysByRead = _interopRequireDefault(require("./useActivityKeysByRead"));

var _useAvatarForBot = _interopRequireDefault(require("./useAvatarForBot"));

var _useAvatarForUser = _interopRequireDefault(require("./useAvatarForUser"));

var _useByteFormatter = _interopRequireDefault(require("./useByteFormatter"));

var _useConnectivityStatus = _interopRequireDefault(require("./useConnectivityStatus"));

var _useCreateActivityRenderer = _interopRequireDefault(require("./useCreateActivityRenderer"));

var _useCreateActivityStatusRenderer = _interopRequireDefault(require("./useCreateActivityStatusRenderer"));

var _useCreateAttachmentForScreenReaderRenderer = _interopRequireDefault(require("./useCreateAttachmentForScreenReaderRenderer"));

var _useCreateAvatarRenderer = _interopRequireDefault(require("./useCreateAvatarRenderer"));

var _useCreateScrollToEndButtonRenderer = _interopRequireDefault(require("./useCreateScrollToEndButtonRenderer"));

var _useDateFormatter = _interopRequireDefault(require("./useDateFormatter"));

var _useDebouncedNotifications = _interopRequireDefault(require("./useDebouncedNotifications"));

var _useDictateInterims = _interopRequireDefault(require("./useDictateInterims"));

var _useDictateState = _interopRequireDefault(require("./useDictateState"));

var _useDirection = _interopRequireDefault(require("./useDirection"));

var _useDisabled = _interopRequireDefault(require("./useDisabled"));

var _useDismissNotification = _interopRequireDefault(require("./useDismissNotification"));

var _useEmitTypingIndicator = _interopRequireDefault(require("./useEmitTypingIndicator"));

var _useGetActivityByKey = _interopRequireDefault(require("./useGetActivityByKey"));

var _useGetHasAcknowledgedByActivityKey = _interopRequireDefault(require("./useGetHasAcknowledgedByActivityKey"));

var _useGetKeyByActivity = _interopRequireDefault(require("./useGetKeyByActivity"));

var _useGetKeyByActivityId = _interopRequireDefault(require("./useGetKeyByActivityId"));

var _useGetSendTimeoutForActivity = _interopRequireDefault(require("./useGetSendTimeoutForActivity"));

var _useGrammars = _interopRequireDefault(require("./useGrammars"));

var _useGroupActivities = _interopRequireDefault(require("./useGroupActivities"));

var _useGroupTimestamp = _interopRequireDefault(require("./useGroupTimestamp"));

var _useLanguage = _interopRequireDefault(require("./useLanguage"));

var _useLastAcknowledgedActivityKey = _interopRequireDefault(require("./useLastAcknowledgedActivityKey"));

var _useLastReadActivityKey = _interopRequireDefault(require("./useLastReadActivityKey"));

var _useLastTypingAt = _interopRequireDefault(require("./useLastTypingAt"));

var _useLocalize = _interopRequireDefault(require("./useLocalize"));

var _useLocalizeDate = _interopRequireDefault(require("./useLocalizeDate"));

var _useLocalizer = _interopRequireDefault(require("./useLocalizer"));

var _useMarkActivityAsSpoken = _interopRequireDefault(require("./useMarkActivityAsSpoken"));

var _useMarkActivityKeyAsRead = _interopRequireDefault(require("./useMarkActivityKeyAsRead"));

var _useMarkAllAsAcknowledged = _interopRequireDefault(require("./useMarkAllAsAcknowledged"));

var _usePerformCardAction = _interopRequireDefault(require("./usePerformCardAction"));

var _usePostActivity = _interopRequireDefault(require("./usePostActivity"));

var _useReferenceGrammarID = _interopRequireDefault(require("./useReferenceGrammarID"));

var _useRelativeTimeFormatter = _interopRequireDefault(require("./useRelativeTimeFormatter"));

var _useRenderActivity = _interopRequireDefault(require("./useRenderActivity"));

var _useRenderActivityStatus = _interopRequireDefault(require("./useRenderActivityStatus"));

var _useRenderAttachment = _interopRequireDefault(require("./useRenderAttachment"));

var _useRenderAvatar = _interopRequireDefault(require("./useRenderAvatar"));

var _useRenderToast = _interopRequireDefault(require("./useRenderToast"));

var _useRenderTypingIndicator = _interopRequireDefault(require("./useRenderTypingIndicator"));

var _useSendBoxValue = _interopRequireDefault(require("./useSendBoxValue"));

var _useSendEvent = _interopRequireDefault(require("./useSendEvent"));

var _useSendFiles = _interopRequireDefault(require("./useSendFiles"));

var _useSendMessage = _interopRequireDefault(require("./useSendMessage"));

var _useSendMessageBack = _interopRequireDefault(require("./useSendMessageBack"));

var _useSendPostBack = _interopRequireDefault(require("./useSendPostBack"));

var _useSendTimeoutForActivity = _interopRequireDefault(require("./useSendTimeoutForActivity"));

var _useSendTypingIndicator = _interopRequireDefault(require("./useSendTypingIndicator"));

var _useSetNotification = _interopRequireDefault(require("./useSetNotification"));

var _useShouldSpeakIncomingActivity = _interopRequireDefault(require("./useShouldSpeakIncomingActivity"));

var _useStartDictate = _interopRequireDefault(require("./useStartDictate"));

var _useStopDictate = _interopRequireDefault(require("./useStopDictate"));

var _useStyleOptions = _interopRequireDefault(require("./useStyleOptions"));

var _useSubmitSendBox = _interopRequireDefault(require("./useSubmitSendBox"));

var _useSuggestedActions = _interopRequireDefault(require("./useSuggestedActions"));

var _useTimeoutForSend = _interopRequireDefault(require("./useTimeoutForSend"));

var _useTrackDimension = _interopRequireDefault(require("./useTrackDimension"));

var _useTrackEvent = _interopRequireDefault(require("./useTrackEvent"));

var _useTrackException = _interopRequireDefault(require("./useTrackException"));

var _useTrackTiming = _interopRequireDefault(require("./useTrackTiming"));

var _useUserID = _interopRequireDefault(require("./useUserID"));

var _useUsername = _interopRequireDefault(require("./useUsername"));

var _useVoiceSelector = _interopRequireDefault(require("./useVoiceSelector"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
