// botframework-webchat-component/lib/ScreenReaderActivity.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _botframeworkWebchatApi = require("botframework-webchat-api");

var _classnames = _interopRequireDefault(require("classnames"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireWildcard(require("react"));

var _activityAltText = _interopRequireDefault(require("./Utils/activityAltText"));

var _useStyleToEmotionObject = _interopRequireDefault(require("./hooks/internal/useStyleToEmotionObject"));

var _useRenderMarkdownAsHTML = _interopRequireDefault(require("./hooks/useRenderMarkdownAsHTML"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

var useAvatarForBot = _botframeworkWebchatApi.hooks.useAvatarForBot,
    useCreateAttachmentForScreenReaderRenderer = _botframeworkWebchatApi.hooks.useCreateAttachmentForScreenReaderRenderer,
    useDateFormatter = _botframeworkWebchatApi.hooks.useDateFormatter,
    useGetKeyByActivity = _botframeworkWebchatApi.hooks.useGetKeyByActivity,
    useLocalizer = _botframeworkWebchatApi.hooks.useLocalizer;
var ROOT_STYLE = {
  '&.webchat__screen-reader-activity': {
    color: 'transparent',
    height: 1,
    opacity: 0,
    overflow: 'hidden',
    position: 'absolute',
    top: 0,
    whiteSpace: 'nowrap',
    width: 1
  }
};
var ACTIVITY_NUM_ATTACHMENTS_ALT_IDS = {
  few: 'ACTIVITY_NUM_ATTACHMENTS_FEW_ALT',
  many: 'ACTIVITY_NUM_ATTACHMENTS_MANY_ALT',
  one: 'ACTIVITY_NUM_ATTACHMENTS_ONE_ALT',
  other: 'ACTIVITY_NUM_ATTACHMENTS_OTHER_ALT',
  two: 'ACTIVITY_NUM_ATTACHMENTS_TWO_ALT'
};

var ScreenReaderAttachments = function ScreenReaderAttachments(_ref) {
  var activity = _ref.activity,
      renderAttachments = _ref.renderAttachments;
  var _activity$attachments = activity.attachments,
      attachments = _activity$attachments === void 0 ? [] : _activity$attachments;
  var createAttachmentForScreenReaderRenderer = useCreateAttachmentForScreenReaderRenderer();
  var localizeWithPlural = useLocalizer({
    plural: true
  });
  var attachmentForScreenReaderRenderers = renderAttachments ? attachments.map(function (attachment) {
    return createAttachmentForScreenReaderRenderer({
      activity: activity,
      attachment: attachment
    });
  }).filter(function (render) {
    return render;
  }) : [];
  var numGenericAttachments = attachments.length - attachmentForScreenReaderRenderers.length;
  var numAttachmentsAlt = !!numGenericAttachments && localizeWithPlural(ACTIVITY_NUM_ATTACHMENTS_ALT_IDS, numGenericAttachments);
  return /*#__PURE__*/_react.default.createElement(_react.Fragment, null, !!attachmentForScreenReaderRenderers.length && /*#__PURE__*/_react.default.createElement("ul", null, attachmentForScreenReaderRenderers.map(function (render, index) {
    return (
      /*#__PURE__*/
      // eslint-disable-next-line react/no-array-index-key
      _react.default.createElement("li", {
        key: index
      }, render())
    );
  })), numAttachmentsAlt && /*#__PURE__*/_react.default.createElement("p", null, numAttachmentsAlt));
};

ScreenReaderAttachments.propTypes = {
  activity: _propTypes.default.shape({
    attachments: _propTypes.default.array
  }).isRequired,
  renderAttachments: _propTypes.default.bool.isRequired
}; // When "renderAttachments" is false, we will not render the content of attachments.
// That means, it will only render "2 attachments", instead of "image attachment".
// This is used in the visual transcript, where we render "Press ENTER to interact."

var ScreenReaderActivity = function ScreenReaderActivity(_ref2) {
  var activity = _ref2.activity,
      children = _ref2.children,
      id = _ref2.id,
      renderAttachments = _ref2.renderAttachments;

  var _useAvatarForBot = useAvatarForBot(),
      _useAvatarForBot2 = _slicedToArray(_useAvatarForBot, 1),
      botInitials = _useAvatarForBot2[0].initials;

  var formatDate = useDateFormatter();
  var getKeyByActivity = useGetKeyByActivity();
  var localize = useLocalizer();
  var renderMarkdownAsHTML = (0, _useRenderMarkdownAsHTML.default)();
  var rootClassName = (0, _useStyleToEmotionObject.default)()(ROOT_STYLE) + '';
  var textAlt = (0, _react.useMemo)(function () {
    return (0, _activityAltText.default)(activity, renderMarkdownAsHTML);
  }, [activity, renderMarkdownAsHTML]);
  var _activity$channelData = activity.channelData;
  _activity$channelData = _activity$channelData === void 0 ? {} : _activity$channelData;
  var fallbackText = _activity$channelData['webchat:fallback-text'],
      _activity$from = activity.from;
  _activity$from = _activity$from === void 0 ? {} : _activity$from;
  var role = _activity$from.role,
      timestamp = activity.timestamp;
  var fromUser = role === 'user';
  var greetingAlt = (fromUser ? localize('ACTIVITY_YOU_SAID_ALT') : localize('ACTIVITY_BOT_SAID_ALT', botInitials || '')).replace(/[\t-\r \xA0\u1680\u2000-\u200A\u2028\u2029\u202F\u205F\u3000\uFEFF]{2,}/g, ' ');
  var timestampAlt = localize('ACTIVITY_STATUS_SEND_STATUS_ALT_SENT_AT', formatDate(timestamp));
  var labelId = (0, _react.useMemo)(function () {
    return "webchat__screen-reader-activity__label-".concat(getKeyByActivity(activity));
  }, [activity, getKeyByActivity]);
  return /*#__PURE__*/_react.default.createElement("article", {
    "aria-atomic": true // Narrator requires the "aria-labelledby" attribute, otherwise, it will only read "aria-roledescription".
    // However, iOS VoiceOver and NVDA both ignore the "aria-labelledby" and read out the whole content, including timestamp.
    ,
    "aria-labelledby": labelId,
    "aria-roledescription": "message",
    className: (0, _classnames.default)('webchat__screen-reader-activity', rootClassName) // "id" attribute is used by `aria-labelledby`.
    // eslint-disable-next-line react/forbid-dom-props
    ,
    id: id
  }, /*#__PURE__*/_react.default.createElement("p", {
    id: labelId
  }, /*#__PURE__*/_react.default.createElement("span", null, greetingAlt), /*#__PURE__*/_react.default.createElement("span", null, textAlt)), !fallbackText && /*#__PURE__*/_react.default.createElement(ScreenReaderAttachments, {
    activity: activity,
    renderAttachments: renderAttachments
  }), /*#__PURE__*/_react.default.createElement("p", {
    className: "webchat__screen-reader-activity__timestamp"
  }, timestampAlt), children);
};

ScreenReaderActivity.defaultProps = {
  children: undefined,
  id: undefined,
  renderAttachments: true
};
ScreenReaderActivity.propTypes = {
  activity: _propTypes.default.any.isRequired,
  children: _propTypes.default.any,
  id: _propTypes.default.string,
  renderAttachments: _propTypes.default.bool
};
var _default = ScreenReaderActivity;
exports.default = _default;
