// react-dictate-button/lib/index.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

_Object$defineProperty(exports, "Composer", {
  enumerable: true,
  get: function get() {
    return _Composer["default"];
  }
});

_Object$defineProperty(exports, "Context", {
  enumerable: true,
  get: function get() {
    return _Context["default"];
  }
});

_Object$defineProperty(exports, "DictateCheckbox", {
  enumerable: true,
  get: function get() {
    return _DictateCheckbox["default"];
  }
});

_Object$defineProperty(exports, "useAbortable", {
  enumerable: true,
  get: function get() {
    return _useAbortable["default"];
  }
});

_Object$defineProperty(exports, "useReadyState", {
  enumerable: true,
  get: function get() {
    return _useReadyState["default"];
  }
});

_Object$defineProperty(exports, "useSupported", {
  enumerable: true,
  get: function get() {
    return _useSupported["default"];
  }
});

exports["default"] = void 0;

var _Composer = _interopRequireDefault(require("./Composer"));

var _Context = _interopRequireDefault(require("./Context"));

var _DictateButton = _interopRequireDefault(require("./DictateButton"));

var _DictateCheckbox = _interopRequireDefault(require("./DictateCheckbox"));

var _useAbortable = _interopRequireDefault(require("./hooks/useAbortable"));

var _useReadyState = _interopRequireDefault(require("./hooks/useReadyState"));

var _useSupported = _interopRequireDefault(require("./hooks/useSupported"));

var _default = _DictateButton["default"];
exports["default"] = _default;
