// botframework-webchat-core/lib/sagas/clearSuggestedActionsOnPostActivitySaga.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = clearSuggestedActionsOnPostActivitySaga;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _effects = require("redux-saga/effects");

var _postActivity = require("../actions/postActivity");

var _setSuggestedActions = _interopRequireDefault(require("../actions/setSuggestedActions"));

var _whileConnected = _interopRequireDefault(require("./effects/whileConnected"));

var _marked = /*#__PURE__*/_regenerator["default"].mark(clearSuggestedActions),
    _marked2 = /*#__PURE__*/_regenerator["default"].mark(clearSuggestedActionsOnPostActivity),
    _marked3 = /*#__PURE__*/_regenerator["default"].mark(clearSuggestedActionsOnPostActivitySaga);

function clearSuggestedActions() {
  return _regenerator["default"].wrap(function clearSuggestedActions$(_context) {
    while (1) {
      switch (_context.prev = _context.next) {
        case 0:
          _context.next = 2;
          return (0, _effects.put)((0, _setSuggestedActions["default"])());

        case 2:
        case "end":
          return _context.stop();
      }
    }
  }, _marked);
}

function clearSuggestedActionsOnPostActivity() {
  return _regenerator["default"].wrap(function clearSuggestedActionsOnPostActivity$(_context2) {
    while (1) {
      switch (_context2.prev = _context2.next) {
        case 0:
          _context2.next = 2;
          return (0, _effects.takeEvery)(function (_ref) {
            var payload = _ref.payload,
                type = _ref.type;
            return type === _postActivity.POST_ACTIVITY_PENDING && payload.activity.type === 'message';
          }, clearSuggestedActions);

        case 2:
        case "end":
          return _context2.stop();
      }
    }
  }, _marked2);
}

function clearSuggestedActionsOnPostActivitySaga() {
  return _regenerator["default"].wrap(function clearSuggestedActionsOnPostActivitySaga$(_context3) {
    while (1) {
      switch (_context3.prev = _context3.next) {
        case 0:
          _context3.next = 2;
          return (0, _whileConnected["default"])(clearSuggestedActionsOnPostActivity);

        case 2:
        case "end":
          return _context3.stop();
      }
    }
  }, _marked3);
}
