// botframework-webchat-core/lib/sagas/emitTypingIndicatorToPostActivitySaga.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = emitTypingActivityToPostActivitySaga;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _effects = require("redux-saga/effects");

var _emitTypingIndicator = require("../actions/emitTypingIndicator");

var _postActivity = _interopRequireDefault(require("../actions/postActivity"));

var _whileConnected = _interopRequireDefault(require("./effects/whileConnected"));

var _marked = /*#__PURE__*/_regenerator["default"].mark(postTypingActivity),
    _marked2 = /*#__PURE__*/_regenerator["default"].mark(emitTypingActivityToPostActivity),
    _marked3 = /*#__PURE__*/_regenerator["default"].mark(emitTypingActivityToPostActivitySaga);

function postTypingActivity() {
  return _regenerator["default"].wrap(function postTypingActivity$(_context) {
    while (1) {
      switch (_context.prev = _context.next) {
        case 0:
          _context.next = 2;
          return (0, _effects.put)((0, _postActivity["default"])({
            type: 'typing'
          }));

        case 2:
        case "end":
          return _context.stop();
      }
    }
  }, _marked);
}

function emitTypingActivityToPostActivity() {
  return _regenerator["default"].wrap(function emitTypingActivityToPostActivity$(_context2) {
    while (1) {
      switch (_context2.prev = _context2.next) {
        case 0:
          _context2.next = 2;
          return (0, _effects.takeEvery)(function (_ref) {
            var type = _ref.type;
            return type === _emitTypingIndicator.EMIT_TYPING_INDICATOR;
          }, postTypingActivity);

        case 2:
        case "end":
          return _context2.stop();
      }
    }
  }, _marked2);
}

function emitTypingActivityToPostActivitySaga() {
  return _regenerator["default"].wrap(function emitTypingActivityToPostActivitySaga$(_context3) {
    while (1) {
      switch (_context3.prev = _context3.next) {
        case 0:
          _context3.next = 2;
          return (0, _whileConnected["default"])(emitTypingActivityToPostActivity);

        case 2:
        case "end":
          return _context3.stop();
      }
    }
  }, _marked3);
}
