// botframework-webchat-core/lib/sagas/sendEventToPostActivitySaga.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = sendEventToPostActivitySaga;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _effects = require("redux-saga/effects");

var _sendEvent = require("../actions/sendEvent");

var _postActivity = _interopRequireDefault(require("../actions/postActivity"));

var _whileConnected = _interopRequireDefault(require("./effects/whileConnected"));

var _marked = /*#__PURE__*/_regenerator["default"].mark(postActivityWithEvent),
    _marked2 = /*#__PURE__*/_regenerator["default"].mark(sendEventToPostActivity),
    _marked3 = /*#__PURE__*/_regenerator["default"].mark(sendEventToPostActivitySaga);

function postActivityWithEvent(_ref) {
  var _ref$payload, name, value;

  return _regenerator["default"].wrap(function postActivityWithEvent$(_context) {
    while (1) {
      switch (_context.prev = _context.next) {
        case 0:
          _ref$payload = _ref.payload, name = _ref$payload.name, value = _ref$payload.value;
          _context.next = 3;
          return (0, _effects.put)((0, _postActivity["default"])({
            name: name,
            type: 'event',
            value: value
          }));

        case 3:
        case "end":
          return _context.stop();
      }
    }
  }, _marked);
}

function sendEventToPostActivity() {
  return _regenerator["default"].wrap(function sendEventToPostActivity$(_context2) {
    while (1) {
      switch (_context2.prev = _context2.next) {
        case 0:
          _context2.next = 2;
          return (0, _effects.takeEvery)(function (_ref2) {
            var payload = _ref2.payload,
                type = _ref2.type;
            return type === _sendEvent.SEND_EVENT && payload.name;
          }, postActivityWithEvent);

        case 2:
        case "end":
          return _context2.stop();
      }
    }
  }, _marked2);
}

function sendEventToPostActivitySaga() {
  return _regenerator["default"].wrap(function sendEventToPostActivitySaga$(_context3) {
    while (1) {
      switch (_context3.prev = _context3.next) {
        case 0:
          _context3.next = 2;
          return (0, _whileConnected["default"])(sendEventToPostActivity);

        case 2:
        case "end":
          return _context3.stop();
      }
    }
  }, _marked3);
}
