// botframework-webchat/lib/adaptiveCards/Attachment/AnimationCardAttachment.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _AnimationCardContent = _interopRequireDefault(require("./AnimationCardContent"));

var AnimationCardAttachment = function AnimationCardAttachment(_ref) {
  var content = _ref.attachment.content,
      disabled = _ref.disabled;
  return /*#__PURE__*/_react.default.createElement(_AnimationCardContent.default, {
    content: content,
    disabled: disabled
  });
};

AnimationCardAttachment.defaultProps = {
  disabled: undefined
};
AnimationCardAttachment.propTypes = {
  attachment: _propTypes.default.shape({
    content: _propTypes.default.shape({
      media: _propTypes.default.arrayOf(_propTypes.default.shape({
        profile: _propTypes.default.string,
        url: _propTypes.default.string.isRequired
      })).isRequired
    }).isRequired
  }).isRequired,
  disabled: _propTypes.default.bool
};
var _default = AnimationCardAttachment;
exports.default = _default;
