// botframework-webchat/lib/adaptiveCards/Attachment/AudioCardAttachment.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _AudioCardContent = _interopRequireDefault(require("./AudioCardContent"));

/* eslint react/no-array-index-key: "off" */
var AudioCardAttachment = function AudioCardAttachment(_ref) {
  var content = _ref.attachment.content,
      disabled = _ref.disabled;
  return /*#__PURE__*/_react.default.createElement(_AudioCardContent.default, {
    content: content,
    disabled: disabled
  });
};

AudioCardAttachment.defaultProps = {
  disabled: undefined
};
AudioCardAttachment.propTypes = {
  attachment: _propTypes.default.shape({
    content: _propTypes.default.shape({
      autostart: _propTypes.default.bool,
      autoloop: _propTypes.default.bool,
      image: _propTypes.default.shape({
        url: _propTypes.default.string.isRequired
      }),
      media: _propTypes.default.arrayOf(_propTypes.default.shape({
        url: _propTypes.default.string.isRequired
      }).isRequired).isRequired
    })
  }).isRequired,
  disabled: _propTypes.default.bool
};
var _default = AudioCardAttachment;
exports.default = _default;
