// botframework-webchat/lib/adaptiveCards/Attachment/HeroCardAttachment.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _HeroCardContent = _interopRequireDefault(require("./HeroCardContent"));

var HeroCardAttachment = function HeroCardAttachment(_ref) {
  var _ref$attachment = _ref.attachment;
  _ref$attachment = _ref$attachment === void 0 ? {} : _ref$attachment;
  var content = _ref$attachment.content,
      disabled = _ref.disabled;
  return !!content && /*#__PURE__*/_react.default.createElement(_HeroCardContent.default, {
    content: content,
    disabled: disabled
  });
};

HeroCardAttachment.defaultProps = {
  disabled: undefined
};
HeroCardAttachment.propTypes = {
  attachment: _propTypes.default.shape({
    content: _propTypes.default.shape({
      buttons: _propTypes.default.any,
      images: _propTypes.default.arrayOf(_propTypes.default.shape({
        alt: _propTypes.default.string,
        tap: _propTypes.default.any,
        url: _propTypes.default.string.isRequired
      })),
      subtitle: _propTypes.default.string,
      tap: _propTypes.default.any,
      text: _propTypes.default.string,
      title: _propTypes.default.string
    }).isRequired
  }).isRequired,
  disabled: _propTypes.default.bool
};
var _default = HeroCardAttachment;
exports.default = _default;
