// botframework-webchat/lib/adaptiveCards/Attachment/VideoCardAttachment.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _VideoCardContent = _interopRequireDefault(require("./VideoCardContent"));

/* eslint react/no-array-index-key: "off" */
var VideoCardAttachment = function VideoCardAttachment(_ref) {
  var content = _ref.attachment.content,
      disabled = _ref.disabled;
  return /*#__PURE__*/_react.default.createElement(_VideoCardContent.default, {
    content: content,
    disabled: disabled
  });
};

VideoCardAttachment.defaultProps = {
  disabled: undefined
};
VideoCardAttachment.propTypes = {
  attachment: _propTypes.default.shape({
    content: _propTypes.default.shape({
      autoloop: _propTypes.default.bool,
      autostart: _propTypes.default.bool,
      image: _propTypes.default.shape({
        url: _propTypes.default.string.isRequired
      }),
      media: _propTypes.default.arrayOf(_propTypes.default.shape({
        profile: _propTypes.default.string.isRequired,
        url: _propTypes.default.string
      }))
    })
  }).isRequired,
  disabled: _propTypes.default.bool
};
var _default = VideoCardAttachment;
exports.default = _default;
