// botframework-webchat/lib/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  createDirectLine: true,
  createDirectLineAppServiceExtension: true,
  Components: true,
  hooks: true,
  renderWebChat: true,
  createAdaptiveCardsAttachmentMiddleware: true,
  createAdaptiveCardsAttachmentForScreenReaderMiddleware: true,
  createCognitiveServicesSpeechServicesPonyfillFactory: true,
  createDirectLineSpeechAdapters: true,
  createStyleSet: true,
  renderMarkdown: true
};
exports.Components = void 0;
Object.defineProperty(exports, "createAdaptiveCardsAttachmentForScreenReaderMiddleware", {
  enumerable: true,
  get: function get() {
    return _createAdaptiveCardsAttachmentForScreenReaderMiddleware.default;
  }
});
Object.defineProperty(exports, "createAdaptiveCardsAttachmentMiddleware", {
  enumerable: true,
  get: function get() {
    return _createAdaptiveCardsAttachmentMiddleware.default;
  }
});
Object.defineProperty(exports, "createCognitiveServicesSpeechServicesPonyfillFactory", {
  enumerable: true,
  get: function get() {
    return _createCognitiveServicesSpeechServicesPonyfillFactory.default;
  }
});
exports.createDirectLineAppServiceExtension = exports.createDirectLine = void 0;
Object.defineProperty(exports, "createDirectLineSpeechAdapters", {
  enumerable: true,
  get: function get() {
    return _createDirectLineSpeechAdapters.default;
  }
});
Object.defineProperty(exports, "createStyleSet", {
  enumerable: true,
  get: function get() {
    return _createFullStyleSet.default;
  }
});
exports.hooks = exports.default = void 0;
Object.defineProperty(exports, "renderMarkdown", {
  enumerable: true,
  get: function get() {
    return _renderMarkdown.default;
  }
});
exports.renderWebChat = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _indexMinimal = require("./index-minimal");

Object.keys(_indexMinimal).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _indexMinimal[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _indexMinimal[key];
    }
  });
});

var _AdaptiveCardContent = _interopRequireDefault(require("./adaptiveCards/Attachment/AdaptiveCardContent"));

var _addVersion = _interopRequireDefault(require("./addVersion"));

var _AnimationCardContent = _interopRequireDefault(require("./adaptiveCards/Attachment/AnimationCardContent"));

var _AudioCardContent = _interopRequireDefault(require("./adaptiveCards/Attachment/AudioCardContent"));

var _renderWebChat = _interopRequireDefault(require("./renderWebChat"));

var _createAdaptiveCardsAttachmentMiddleware = _interopRequireDefault(require("./adaptiveCards/createAdaptiveCardsAttachmentMiddleware"));

var _createAdaptiveCardsAttachmentForScreenReaderMiddleware = _interopRequireDefault(require("./adaptiveCards/createAdaptiveCardsAttachmentForScreenReaderMiddleware"));

var _createCognitiveServicesSpeechServicesPonyfillFactory = _interopRequireDefault(require("./createCognitiveServicesSpeechServicesPonyfillFactory"));

var _createDirectLineSpeechAdapters = _interopRequireDefault(require("./createDirectLineSpeechAdapters"));

var _createFullStyleSet = _interopRequireDefault(require("./createFullStyleSet"));

var _createDirectLine = _interopRequireDefault(require("./createDirectLine"));

var _createDirectLineAppServiceExtension = _interopRequireDefault(require("./createDirectLineAppServiceExtension"));

var _FullComposer = _interopRequireDefault(require("./FullComposer"));

var _HeroCardContent = _interopRequireDefault(require("./adaptiveCards/Attachment/HeroCardContent"));

var _OAuthCardContent = _interopRequireDefault(require("./adaptiveCards/Attachment/OAuthCardContent"));

var _FullReactWebChat = _interopRequireDefault(require("./FullReactWebChat"));

var _ReceiptCardContent = _interopRequireDefault(require("./adaptiveCards/Attachment/ReceiptCardContent"));

var _renderMarkdown = _interopRequireDefault(require("./renderMarkdown"));

var _SignInCardContent = _interopRequireDefault(require("./adaptiveCards/Attachment/SignInCardContent"));

var _ThumbnailCardContent = _interopRequireDefault(require("./adaptiveCards/Attachment/ThumbnailCardContent"));

var _useAdaptiveCardsHostConfig = _interopRequireDefault(require("./adaptiveCards/hooks/useAdaptiveCardsHostConfig"));

var _useAdaptiveCardsPackage = _interopRequireDefault(require("./adaptiveCards/hooks/useAdaptiveCardsPackage"));

var _useStyleOptions = _interopRequireDefault(require("./hooks/useStyleOptions"));

var _VideoCardContent = _interopRequireDefault(require("./adaptiveCards/Attachment/VideoCardContent"));

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var renderWebChat = _renderWebChat.default.bind(null, _FullReactWebChat.default);

exports.renderWebChat = renderWebChat;

var createDirectLine = function createDirectLine(options) {
  options.botAgent && console.warn('Web Chat: Developers are not currently allowed to set botAgent. See https://github.com/microsoft/BotFramework-WebChat/issues/2119 for more details.');
  return (0, _createDirectLine.default)(_objectSpread(_objectSpread({}, options), {}, {
    botAgent: "WebChat/".concat(_indexMinimal.version, " (Full)")
  }));
};

exports.createDirectLine = createDirectLine;

var createDirectLineAppServiceExtension = function createDirectLineAppServiceExtension(options) {
  options.botAgent && console.warn('Web Chat: Developers are not currently allowed to set botAgent. See https://github.com/microsoft/BotFramework-WebChat/issues/2119 for more details.');
  return (0, _createDirectLineAppServiceExtension.default)(_objectSpread(_objectSpread({}, options), {}, {
    botAgent: "WebChat/".concat(_indexMinimal.version, " (Full)")
  }));
};

exports.createDirectLineAppServiceExtension = createDirectLineAppServiceExtension;

var patchedHooks = _objectSpread(_objectSpread({}, _indexMinimal.hooks), {}, {
  useAdaptiveCardsHostConfig: _useAdaptiveCardsHostConfig.default,
  useAdaptiveCardsPackage: _useAdaptiveCardsPackage.default,
  useStyleOptions: _useStyleOptions.default
});

exports.hooks = patchedHooks;
var AdditionalComponents = {
  AdaptiveCardContent: _AdaptiveCardContent.default,
  AnimationCardContent: _AnimationCardContent.default,
  AudioCardContent: _AudioCardContent.default,
  Composer: _FullComposer.default,
  HeroCardContent: _HeroCardContent.default,
  OAuthCardContent: _OAuthCardContent.default,
  ReceiptCardContent: _ReceiptCardContent.default,
  SignInCardContent: _SignInCardContent.default,
  ThumbnailCardContent: _ThumbnailCardContent.default,
  VideoCardContent: _VideoCardContent.default
};

var Components = _objectSpread(_objectSpread({}, _indexMinimal.Components), AdditionalComponents);

exports.Components = Components;
var _default = _FullReactWebChat.default;
exports.default = _default;
window['WebChat'] = _objectSpread(_objectSpread({}, window['WebChat']), {}, {
  Components: Components,
  createAdaptiveCardsAttachmentMiddleware: _createAdaptiveCardsAttachmentMiddleware.default,
  createAdaptiveCardsAttachmentForScreenReaderMiddleware: _createAdaptiveCardsAttachmentForScreenReaderMiddleware.default,
  createCognitiveServicesSpeechServicesPonyfillFactory: _createCognitiveServicesSpeechServicesPonyfillFactory.default,
  createDirectLine: createDirectLine,
  createDirectLineAppServiceExtension: createDirectLineAppServiceExtension,
  createDirectLineSpeechAdapters: _createDirectLineSpeechAdapters.default,
  createStyleSet: _createFullStyleSet.default,
  hooks: patchedHooks,
  ReactWebChat: _FullReactWebChat.default,
  renderMarkdown: _renderMarkdown.default,
  renderWebChat: renderWebChat
});
(0, _addVersion.default)('full');
