// botframework-webchat/lib/renderMarkdown.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = render;

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _markdownItForInline = _interopRequireDefault(require("markdown-it-for-inline"));

var _markdownIt = _interopRequireDefault(require("markdown-it"));

var _markdownItAttrsEs = _interopRequireDefault(require("markdown-it-attrs-es5"));

var _sanitizeHtml = _interopRequireDefault(require("sanitize-html"));

/* eslint no-magic-numbers: ["error", { "ignore": [0, 1, 2] }] */
var SANITIZE_HTML_OPTIONS = {
  allowedAttributes: {
    a: ['aria-label', 'href', 'name', 'rel', 'target', 'title'],
    img: ['alt', 'class', 'src']
  },
  allowedSchemes: ['data', 'http', 'https', 'ftp', 'mailto', 'sip', 'tel'],
  allowedTags: ['a', 'b', 'blockquote', 'br', 'caption', 'code', 'del', 'div', 'em', 'h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'hr', 'i', 'img', 'ins', 'li', 'nl', 'ol', 'p', 'pre', 's', 'span', 'strike', 'strong', 'table', 'tbody', 'td', 'tfoot', 'th', 'thead', 'tr', 'ul']
}; // Put a transparent pixel instead of the "open in new window" icon, so developers can easily modify the icon in CSS.

var TRANSPARENT_GIF = 'data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7'; // This is used for parsing Markdown for external links.

var internalMarkdownIt = new _markdownIt.default();
var MARKDOWN_ATTRS_LEFT_DELIMITER = '⟬'; // Make sure the delimiter is free from any RegExp characters, such as *, ?, etc.
// IE11 does not support "u" flag and Babel could not remove it. We intentionally omitting the "u" flag here.
// eslint-disable-next-line security/detect-non-literal-regexp, require-unicode-regexp

var MARKDOWN_ATTRS_LEFT_DELIMITER_PATTERN = new RegExp(MARKDOWN_ATTRS_LEFT_DELIMITER, 'g');
var MARKDOWN_ATTRS_RIGHT_DELIMITER = '⟭'; // Make sure the delimiter is free from any RegExp characters, such as *, ?, etc.
// IE11 does not support "u" flag and Babel could not remove it. We intentionally omitting the "u" flag here.
// eslint-disable-next-line security/detect-non-literal-regexp, require-unicode-regexp

var MARKDOWN_ATTRS_RIGHT_DELIMITER_PATTERN = new RegExp(MARKDOWN_ATTRS_RIGHT_DELIMITER, 'g');

function render(markdown, _ref) {
  var markdownRespectCRLF = _ref.markdownRespectCRLF;

  var _ref2 = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {},
      _ref2$externalLinkAlt = _ref2.externalLinkAlt,
      externalLinkAlt = _ref2$externalLinkAlt === void 0 ? '' : _ref2$externalLinkAlt;

  if (markdownRespectCRLF) {
    markdown = markdown.replace(/\n\r|\r\n/g, function (carriageReturn) {
      return carriageReturn === '\n\r' ? '\r\n' : '\n\r';
    });
  } // Related to #3165.
  // We only support attributes "aria-label" and should leave other attributes as-is.
  // However, `markdown-it-attrs` remove unrecognized attributes, such as {hello}.
  // Before passing to `markdown-it-attrs`, we will convert known attributes from {aria-label="..."} into ⟬aria-label="..."⟭ (using white tortoise shell brackets).
  // Then, we ask `markdown-it-attrs` to only process the new brackets, so it should only try to process things that we allowlisted.
  // Lastly, we revert tortoise shell brackets back to curly brackets, for unprocessed attributes.


  markdown = markdown.replace(/\{[\t-\r \xA0\u1680\u2000-\u200A\u2028\u2029\u202F\u205F\u3000\uFEFF]*aria\x2Dlabel()[\t-\r \xA0\u1680\u2000-\u200A\u2028\u2029\u202F\u205F\u3000\uFEFF]*\}/gi, "".concat(MARKDOWN_ATTRS_LEFT_DELIMITER, "aria-label").concat(MARKDOWN_ATTRS_RIGHT_DELIMITER)).replace(/\{[\t-\r \xA0\u1680\u2000-\u200A\u2028\u2029\u202F\u205F\u3000\uFEFF]*aria\x2Dlabel=("(?:(?!")[\s\S])*"|(?:(?![\t-\r \}\xA0\u1680\u2000-\u200A\u2028\u2029\u202F\u205F\u3000\uFEFF])[\s\S])*)[\t-\r \xA0\u1680\u2000-\u200A\u2028\u2029\u202F\u205F\u3000\uFEFF]*\}/gi, function (_, valueInsideQuotes) {
    return "".concat(MARKDOWN_ATTRS_LEFT_DELIMITER, "aria-label=").concat(valueInsideQuotes).concat(MARKDOWN_ATTRS_RIGHT_DELIMITER);
  });
  var html = new _markdownIt.default({
    breaks: false,
    html: false,
    linkify: true,
    typographer: true,
    xhtmlOut: true
  }).use(_markdownItAttrsEs.default, {
    // `markdown-it-attrs` is added for accessibility and allow bot developers to specify `aria-label`.
    // We are allowlisting `aria-label` only as it is allowlisted in `sanitize-html`.
    // Other `aria-*` will be sanitized even we allowlisted here.
    allowedAttributes: ['aria-label'],
    leftDelimiter: MARKDOWN_ATTRS_LEFT_DELIMITER,
    rightDelimiter: MARKDOWN_ATTRS_RIGHT_DELIMITER
  }).use(_markdownItForInline.default, 'url_new_win', 'link_open', function (tokens, index) {
    var token = tokens[+index];
    token.attrSet('rel', 'noopener noreferrer');
    token.attrSet('target', '_blank');
    var linkOpenToken = tokens.find(function (_ref3) {
      var type = _ref3.type;
      return type === 'link_open';
    });

    var _linkOpenToken$attrs$ = linkOpenToken.attrs.find(function (_ref4) {
      var _ref5 = (0, _slicedToArray2.default)(_ref4, 1),
          name = _ref5[0];

      return name === 'href';
    }),
        _linkOpenToken$attrs$2 = (0, _slicedToArray2.default)(_linkOpenToken$attrs$, 2),
        href = _linkOpenToken$attrs$2[1]; // Adds a new icon if the link is http: or https:.
    // Don't add if it's a phone number, etc.


    if (/^http[s\u017F]?:/i.test(href)) {
      externalLinkAlt && token.attrSet('title', externalLinkAlt);
      var iconTokens = internalMarkdownIt.parseInline("![".concat(externalLinkAlt, "](").concat(TRANSPARENT_GIF, ")"))[0].children;
      iconTokens[0].attrJoin('class', 'webchat__markdown__external-link-icon');
      tokens.splice.apply(tokens, [index + 2, 0].concat((0, _toConsumableArray2.default)(iconTokens)));
    }
  }).render(markdown); // Restore attributes not processed by `markdown-it-attrs`.
  // TODO: [P2] #2511 After we fixed our polyfill story, we should use "String.prototype.replaceAll" instead of RegExp for replace all occurrences.

  html = html.replace(MARKDOWN_ATTRS_LEFT_DELIMITER_PATTERN, '{').replace(MARKDOWN_ATTRS_RIGHT_DELIMITER_PATTERN, '}'); // The signature from "sanitize-html" module is not correct.
  // @ts-ignore

  return (0, _sanitizeHtml.default)(html, SANITIZE_HTML_OPTIONS);
}
