// react-film/lib/esm/ReactFilm.js
import "core-js/modules/es.object.keys.js";
import "core-js/modules/es.symbol.js";
import "core-js/modules/es.array.filter.js";
import "core-js/modules/es.object.get-own-property-descriptor.js";
import "core-js/modules/es.array.for-each.js";
import "core-js/modules/web.dom-collections.for-each.js";
import "core-js/modules/es.object.get-own-property-descriptors.js";
import "core-js/modules/es.object.define-properties.js";
import "core-js/modules/es.object.define-property.js";
import _defineProperty from "@babel/runtime/helpers/defineProperty";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

import PropTypes from 'prop-types';
import React, { Children, useMemo } from 'react';
import BasicFilm from './BasicFilm';
import Composer from './Composer';

var ReactFilm = function ReactFilm(_ref) {
  var autoCenter = _ref.autoCenter,
      autoHide = _ref.autoHide,
      autoHideFlipperOnEdge = _ref.autoHideFlipperOnEdge,
      children = _ref.children,
      className = _ref.className,
      dir = _ref.dir,
      flipperBlurFocusOnClick = _ref.flipperBlurFocusOnClick,
      height = _ref.height,
      leftFlipperText = _ref.leftFlipperText,
      nonce = _ref.nonce,
      rightFlipperText = _ref.rightFlipperText,
      showDots = _ref.showDots,
      showFlipper = _ref.showFlipper,
      showScrollBar = _ref.showScrollBar,
      styleSet = _ref.styleSet;
  var styleOptions = useMemo(function () {
    return {
      autoCenter: autoCenter,
      autoHide: autoHide,
      autoHideFlipperOnEdge: autoHideFlipperOnEdge,
      dir: dir,
      leftFlipperText: leftFlipperText,
      flipperBlurFocusOnClick: flipperBlurFocusOnClick,
      rightFlipperText: rightFlipperText,
      showDots: showDots,
      showFlipper: showFlipper,
      showScrollBar: showScrollBar
    };
  }, [autoCenter, autoHide, autoHideFlipperOnEdge, dir, leftFlipperText, flipperBlurFocusOnClick, rightFlipperText, showDots, showFlipper, showScrollBar]);
  return /*#__PURE__*/React.createElement(Composer, {
    dir: dir,
    height: height,
    nonce: nonce,
    numItems: children ? Children.count(children) : 0,
    styleOptions: styleOptions,
    styleSet: styleSet
  }, /*#__PURE__*/React.createElement(BasicFilm, {
    className: className
  }, children));
};

ReactFilm.defaultProps = _objectSpread(_objectSpread({}, BasicFilm.defaultProps), {}, {
  autoCenter: undefined,
  autoHide: undefined,
  autoHideFlipperOnEdge: undefined,
  children: undefined,
  className: undefined,
  dir: undefined,
  flipperBlurFocusOnClick: undefined,
  leftFlipperText: undefined,
  nonce: undefined,
  rightFlipperText: undefined,
  showDots: undefined,
  showFlipper: undefined,
  showScrollBar: undefined,
  styleSet: undefined
});
ReactFilm.propTypes = _objectSpread(_objectSpread({}, BasicFilm.propTypes), {}, {
  autoCenter: PropTypes.bool,
  autoHide: PropTypes.bool,
  autoHideFlipperOnEdge: PropTypes.bool,
  children: PropTypes.any,
  className: PropTypes.string,
  dir: PropTypes.oneOf(['auto', 'ltr', 'rtl']),
  flipperBlurFocusOnClick: PropTypes.bool,
  height: PropTypes.number,
  leftFlipperText: PropTypes.string,
  nonce: PropTypes.string,
  rightFlipperText: PropTypes.string,
  showDots: PropTypes.bool,
  showFlipper: PropTypes.bool,
  showScrollBar: PropTypes.bool,
  styleSet: PropTypes.any
});
export default ReactFilm;
