// react-film/lib/esm/ScrollBar.js
import "core-js/modules/es.object.keys.js";
import "core-js/modules/es.symbol.js";
import "core-js/modules/es.array.filter.js";
import "core-js/modules/es.object.get-own-property-descriptor.js";
import "core-js/modules/es.array.for-each.js";
import "core-js/modules/web.dom-collections.for-each.js";
import "core-js/modules/es.object.get-own-property-descriptors.js";
import "core-js/modules/es.object.define-properties.js";
import "core-js/modules/es.object.define-property.js";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _slicedToArray from "@babel/runtime/helpers/slicedToArray";
import "core-js/modules/es.parse-float.js";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

/* eslint no-magic-numbers: ["error", { "ignore": [1, 100] }] */
import React from 'react';
import * as browser from './browser';
import useDir from './hooks/useDir';
import useScrollBarPercentage from './hooks/useScrollBarPercentage';
import useScrollBarWidth from './hooks/useScrollBarWidth';

var ScrollBar = function ScrollBar() {
  var _useDir = useDir(),
      _useDir2 = _slicedToArray(_useDir, 1),
      dir = _useDir2[0];

  var _useScrollBarPercenta = useScrollBarPercentage(),
      _useScrollBarPercenta2 = _slicedToArray(_useScrollBarPercenta, 1),
      scrollBarPercentage = _useScrollBarPercenta2[0];

  var _useScrollBarWidth = useScrollBarWidth(),
      _useScrollBarWidth2 = _slicedToArray(_useScrollBarWidth, 1),
      scrollBarWidth = _useScrollBarWidth2[0];

  return /*#__PURE__*/React.createElement("div", {
    className: "react-film__scroll-bar react-film__main__slider react-film__main__slider--bottom"
  }, /*#__PURE__*/React.createElement("div", {
    className: "react-film__scroll-bar__handle",
    style: _objectSpread(_objectSpread({}, dir === 'rtl' ? browser.edgeUWP || browser.internetExplorer ? {
      marginRight: "".concat((1 - parseFloat(scrollBarWidth) / 100) * parseFloat(scrollBarPercentage), "%")
    } : {
      marginRight: "".concat((parseFloat(scrollBarWidth) / 100 - 1) * parseFloat(scrollBarPercentage), "%")
    } : {
      marginLeft: "".concat((1 - parseFloat(scrollBarWidth) / 100) * parseFloat(scrollBarPercentage), "%")
    }), {}, {
      width: scrollBarWidth
    })
  }));
};

export default ScrollBar;
