// react-film/lib/BasicFilm.js
"use strict";

require("core-js/modules/es.array.iterator.js");

require("core-js/modules/es.object.to-string.js");

require("core-js/modules/es.string.iterator.js");

require("core-js/modules/es.weak-map.js");

require("core-js/modules/web.dom-collections.iterator.js");

require("core-js/modules/es.object.define-property.js");

require("core-js/modules/es.object.get-own-property-descriptor.js");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _classnames = _interopRequireDefault(require("classnames"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireWildcard(require("react"));

var _Dots = _interopRequireDefault(require("./Dots"));

var _Filmstrip = _interopRequireDefault(require("./Filmstrip"));

var _Flipper = _interopRequireDefault(require("./Flipper"));

var _ScrollBar = _interopRequireDefault(require("./ScrollBar"));

var _useDir3 = _interopRequireDefault(require("./hooks/useDir"));

var _useHeight3 = _interopRequireDefault(require("./hooks/useHeight"));

var _useNumItems3 = _interopRequireDefault(require("./hooks/useNumItems"));

var _useScrollBarWidth3 = _interopRequireDefault(require("./hooks/useScrollBarWidth"));

var _useScrolling3 = _interopRequireDefault(require("./hooks/useScrolling"));

var _useStyleOptions3 = _interopRequireDefault(require("./hooks/useStyleOptions"));

var _useStyleSetClassNames = _interopRequireDefault(require("./hooks/useStyleSetClassNames"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var BasicFilm = function BasicFilm(_ref) {
  var children = _ref.children,
      className = _ref.className;

  var _useDir = (0, _useDir3["default"])(),
      _useDir2 = (0, _slicedToArray2["default"])(_useDir, 1),
      dir = _useDir2[0];

  var _useHeight = (0, _useHeight3["default"])(),
      _useHeight2 = (0, _slicedToArray2["default"])(_useHeight, 1),
      height = _useHeight2[0];

  var _useNumItems = (0, _useNumItems3["default"])(),
      _useNumItems2 = (0, _slicedToArray2["default"])(_useNumItems, 1),
      numItems = _useNumItems2[0];

  var _useScrollBarWidth = (0, _useScrollBarWidth3["default"])(),
      _useScrollBarWidth2 = (0, _slicedToArray2["default"])(_useScrollBarWidth, 1),
      scrollBarWidth = _useScrollBarWidth2[0];

  var _useScrolling = (0, _useScrolling3["default"])(),
      _useScrolling2 = (0, _slicedToArray2["default"])(_useScrolling, 1),
      scrolling = _useScrolling2[0];

  var _useStyleSetClassName = (0, _useStyleSetClassNames["default"])(),
      _useStyleSetClassName2 = (0, _slicedToArray2["default"])(_useStyleSetClassName, 1),
      rootClassName = _useStyleSetClassName2[0].root;

  var _useStyleOptions = (0, _useStyleOptions3["default"])(),
      _useStyleOptions2 = (0, _slicedToArray2["default"])(_useStyleOptions, 1),
      _useStyleOptions2$ = _useStyleOptions2[0],
      flipperBlurFocusOnClick = _useStyleOptions2$.flipperBlurFocusOnClick,
      leftFlipperText = _useStyleOptions2$.leftFlipperText,
      rightFlipperText = _useStyleOptions2$.rightFlipperText,
      showDots = _useStyleOptions2$.showDots,
      showFlipper = _useStyleOptions2$.showFlipper,
      showScrollBar = _useStyleOptions2$.showScrollBar;

  var contentStyle = (0, _react.useMemo)(function () {
    return {
      height: height
    };
  }, [height]);
  return /*#__PURE__*/_react["default"].createElement("div", {
    className: (0, _classnames["default"])(rootClassName, (className || '') + ''),
    dir: dir
  }, /*#__PURE__*/_react["default"].createElement("div", {
    className: (0, _classnames["default"])('react-film__main', {
      'react-film__main--scrolling': scrolling
    }),
    style: contentStyle
  }, !!numItems && scrollBarWidth !== '100%' && !!showFlipper && /*#__PURE__*/_react["default"].createElement(_Flipper["default"], {
    blurFocusOnClick: flipperBlurFocusOnClick,
    mode: "left"
  }, leftFlipperText), /*#__PURE__*/_react["default"].createElement(_Filmstrip["default"], null, children), !!numItems && scrollBarWidth !== '100%' && !!showFlipper && /*#__PURE__*/_react["default"].createElement(_Flipper["default"], {
    blurFocusOnClick: flipperBlurFocusOnClick,
    mode: "right"
  }, rightFlipperText), !!numItems && scrollBarWidth !== '100%' && !!showScrollBar && /*#__PURE__*/_react["default"].createElement(_ScrollBar["default"], null)), !!numItems && scrollBarWidth !== '100%' && !!showDots && /*#__PURE__*/_react["default"].createElement(_Dots["default"], null));
}; // TODO: Move from styleSet to styleSheet.


BasicFilm.defaultProps = {
  children: undefined,
  className: undefined
};
BasicFilm.propTypes = {
  children: _propTypes["default"].any,
  className: _propTypes["default"].string
};
var _default = BasicFilm;
exports["default"] = _default;
