// react-film/lib/Flipper.js
"use strict";

require("core-js/modules/es.array.iterator.js");

require("core-js/modules/es.object.to-string.js");

require("core-js/modules/es.string.iterator.js");

require("core-js/modules/es.weak-map.js");

require("core-js/modules/web.dom-collections.iterator.js");

require("core-js/modules/es.object.define-property.js");

require("core-js/modules/es.object.get-own-property-descriptor.js");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _classnames = _interopRequireDefault(require("classnames"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireWildcard(require("react"));

var _useDir3 = _interopRequireDefault(require("./hooks/useDir"));

var _useScrollBarPercentage = _interopRequireDefault(require("./hooks/useScrollBarPercentage"));

var _useScrollOneLeft = _interopRequireDefault(require("./hooks/useScrollOneLeft"));

var _useScrollOneRight = _interopRequireDefault(require("./hooks/useScrollOneRight"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var Flipper = function Flipper(_ref) {
  var ariaLabel = _ref['aria-label'],
      blurFocusOnClick = _ref.blurFocusOnClick,
      children = _ref.children,
      mode = _ref.mode;

  var _useDir = (0, _useDir3["default"])(),
      _useDir2 = (0, _slicedToArray2["default"])(_useDir, 1),
      dir = _useDir2[0];

  var _useScrollBarPercenta = (0, _useScrollBarPercentage["default"])(),
      _useScrollBarPercenta2 = (0, _slicedToArray2["default"])(_useScrollBarPercenta, 1),
      scrollBarPercentage = _useScrollBarPercenta2[0];

  var buttonRef = (0, _react.useRef)();
  var left = mode === 'left';
  var scrollOneLeft = (0, _useScrollOneLeft["default"])();
  var scrollOneRight = (0, _useScrollOneRight["default"])();
  var handleClick = (0, _react.useCallback)(function () {
    left ? scrollOneLeft() : scrollOneRight();
    blurFocusOnClick && buttonRef.current && buttonRef.current.blur();
  }, [blurFocusOnClick, buttonRef, left, scrollOneLeft, scrollOneRight]);
  var handleKeyDown = (0, _react.useCallback)(function (event) {
    var key = event.key;

    if (key === 'Enter' || key === ' ') {
      event.preventDefault();
      left ? scrollOneLeft() : scrollOneRight();
    }
  }, [left, scrollOneLeft, scrollOneRight]);
  var hide;

  if (dir === 'rtl') {
    if (left) {
      hide = scrollBarPercentage === '100%' || scrollBarPercentage === '-100%';
    } else {
      hide = scrollBarPercentage === '0%';
    }
  } else {
    if (left) {
      hide = scrollBarPercentage === '0%';
    } else {
      hide = scrollBarPercentage === '100%';
    }
  }

  return /*#__PURE__*/_react["default"].createElement("button", {
    "aria-label": ariaLabel || (left ? 'left' : 'right'),
    className: (0, _classnames["default"])('react-film__flipper', 'react-film__main__overlay', {
      'react-film__flipper--left': left,
      'react-film__flipper--right': !left
    }),
    onClick: handleClick,
    onKeyDown: handleKeyDown,
    ref: buttonRef,
    type: "button"
  }, /*#__PURE__*/_react["default"].createElement("div", {
    className: (0, _classnames["default"])('react-film__flipper__slider', 'react-film__main__slider', {
      'react-film__main__slider--hide': hide,
      'react-film__main__slider--left': left,
      'react-film__main__slider--right': !left
    })
  }, /*#__PURE__*/_react["default"].createElement("div", {
    className: "react-film__flipper__body"
  }, children)));
};

Flipper.defaultProps = {
  'aria-label': undefined,
  blurFocusOnClick: false,
  children: undefined,
  mode: 'left'
};
Flipper.propTypes = {
  'aria-label': _propTypes["default"].string,
  blurFocusOnClick: _propTypes["default"].bool,
  children: _propTypes["default"].any,
  mode: _propTypes["default"].oneOf(['left', 'right'])
};
var _default = Flipper;
exports["default"] = _default;
