// react-dictate-button/lib/esm/Composer.js
import _slicedToArray from "@babel/runtime-corejs3/helpers/slicedToArray";
import _forEachInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/for-each";
import _mapInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/map";
import _Object$entries from "@babel/runtime-corejs3/core-js-stable/object/entries";

/* eslint no-magic-numbers: ["error", { "ignore": [0, 1, 2, 3] }] */
import PropTypes from 'prop-types';
import React, { useCallback, useEffect, useMemo, useRef, useState } from 'react';
import Context from './Context';
import usePrevious from './usePrevious';
import useRefFrom from './useRefFrom';
import vendorPrefix from './vendorPrefix';

function applyAll() {
  for (var _len = arguments.length, fns = new Array(_len), _key = 0; _key < _len; _key++) {
    fns[_key] = arguments[_key];
  }

  return function () {
    var _arguments = arguments,
        _this = this;

    // eslint-disable-next-line no-invalid-this, prefer-rest-params
    _forEachInstanceProperty(fns).call(fns, function (fn) {
      return fn.apply(_this, _arguments);
    });
  };
}

function recognitionAbortable(recognition) {
  return !!(recognition && typeof recognition.abort === 'function');
}

var Composer = function Composer(_ref) {
  var children = _ref.children,
      extra = _ref.extra,
      grammar = _ref.grammar,
      lang = _ref.lang,
      onDictate = _ref.onDictate,
      onError = _ref.onError,
      onProgress = _ref.onProgress,
      onRawEvent = _ref.onRawEvent,
      speechGrammarList = _ref.speechGrammarList,
      speechRecognition = _ref.speechRecognition,
      started = _ref.started;

  var _useState = useState(0),
      _useState2 = _slicedToArray(_useState, 2),
      readyState = _useState2[0],
      setReadyState = _useState2[1];

  var emitDictateOnEndRef = useRef(false);
  var extraRef = useRefFrom(extra);
  var grammarRef = useRefFrom(grammar);
  var langRef = useRefFrom(lang);
  var notAllowedRef = useRef(false);
  var onDictateRef = useRefFrom(onDictate);
  var onErrorRef = useRefFrom(onError);
  var onProgressRef = useRefFrom(onProgress);
  var onRawEventRef = useRefFrom(onRawEvent);
  var prevSpeechRecognition = usePrevious(speechRecognition);
  var recognitionRef = useRef();
  var speechGrammarListRef = useRefFrom(speechGrammarList);
  var speechRecognitionRef = useRefFrom(speechRecognition); // If "speechRecognition" ponyfill changed, reset the "notAllowed" flag.

  if (prevSpeechRecognition !== speechRecognition) {
    notAllowedRef.current = false;
  }

  var handleAudioEnd = useCallback(function (_ref2) {
    var target = _ref2.target;
    return target === recognitionRef.current && setReadyState(3);
  }, [recognitionRef, setReadyState]);
  var handleAudioStart = useCallback(function (_ref3) {
    var target = _ref3.target;

    if (target !== recognitionRef.current) {
      return;
    }

    setReadyState(2); // Web Speech API does not emit "result" when nothing is heard, and Chrome does not emit "nomatch" event.
    // Because we emitted onProgress, we should emit "dictate" if not error, so they works in pair.

    emitDictateOnEndRef.current = true;
    onProgressRef.current && onProgressRef.current({
      abortable: recognitionAbortable(target),
      type: 'progress'
    });
  }, [emitDictateOnEndRef, onProgressRef, recognitionRef, setReadyState]);
  var handleEnd = useCallback(function (_ref4) {
    var target = _ref4.target;

    if (target !== recognitionRef.current) {
      return;
    }

    recognitionRef.current = undefined;
    setReadyState(0);

    if (emitDictateOnEndRef.current) {
      onDictateRef.current && onDictateRef.current({
        type: 'dictate'
      });
      emitDictateOnEndRef.current = false;
    }
  }, [emitDictateOnEndRef, onDictateRef, recognitionRef, setReadyState]);
  var handleError = useCallback(function (event) {
    if (event.target !== recognitionRef.current) {
      return;
    } // Error out, no need to emit "dictate"


    emitDictateOnEndRef.current = false;
    recognitionRef.current = undefined;

    if (event.error === 'not-allowed') {
      notAllowedRef.current = true;
    }

    setReadyState(0);
    onErrorRef.current && onErrorRef.current(event);
  }, [emitDictateOnEndRef, onErrorRef, notAllowedRef, recognitionRef, setReadyState]);
  var handleRawEvent = useCallback(function (event) {
    if (event.target !== recognitionRef.current) {
      return;
    }

    onRawEventRef.current && onRawEventRef.current(event);
  }, [onRawEventRef, recognitionRef]);
  var handleResult = useCallback(function (_ref5) {
    var rawResults = _ref5.results,
        target = _ref5.target;

    if (target !== recognitionRef.current) {
      return;
    }

    if (rawResults.length) {
      var results = _mapInstanceProperty([]).call(rawResults, function (alts) {
        // Destructuring breaks Angular due to a bug in Zone.js.
        // eslint-disable-next-line prefer-destructuring
        var firstAlt = alts[0];
        return {
          confidence: firstAlt.confidence,
          transcript: firstAlt.transcript
        };
      }); // Destructuring breaks Angular due to a bug in Zone.js.
      // eslint-disable-next-line prefer-destructuring


      var first = rawResults[0];

      if (first.isFinal) {
        // After "onDictate" callback, the caller should be able to set "started" to false on an unabortable recognition.
        // TODO: Add test for fortification.
        recognitionRef.current = undefined;
        setReadyState(0);
        onDictateRef.current && onDictateRef.current({
          result: results[0],
          type: 'dictate'
        });
      } else {
        onProgressRef.current && onProgressRef.current({
          abortable: recognitionAbortable(target),
          results: results,
          type: 'progress'
        });
      }
    }
  }, [onDictateRef, onProgressRef, recognitionRef, setReadyState]);
  var handleStart = useCallback(function (_ref6) {
    var target = _ref6.target;
    return target === recognitionRef.current && setReadyState(1);
  }, [recognitionRef, setReadyState]);
  useEffect(function () {
    if (started) {
      var _context;

      if (!speechRecognitionRef.current || notAllowedRef.current) {
        throw new Error('Speech recognition is not supported');
      }

      var grammars = speechGrammarListRef.current && grammarRef.current && new speechGrammarListRef.current();
      var recognition = recognitionRef.current = new speechRecognitionRef.current();

      if (grammars) {
        grammars.addFromString(grammarRef.current, 1);
        recognition.grammars = grammars;
      }

      recognition.lang = langRef.current;
      recognition.interimResults = true;
      recognition.onaudioend = applyAll(handleAudioEnd, handleRawEvent);
      recognition.onaudiostart = applyAll(handleAudioStart, handleRawEvent);
      recognition.onend = applyAll(handleEnd, handleRawEvent);
      recognition.onerror = applyAll(handleError, handleRawEvent);
      recognition.onnomatch = handleRawEvent;
      recognition.onresult = applyAll(handleResult, handleRawEvent);
      recognition.onsoundend = handleRawEvent;
      recognition.onsoundstart = handleRawEvent;
      recognition.onspeechend = handleRawEvent;
      recognition.onspeechstart = handleRawEvent;
      recognition.onstart = applyAll(handleStart, handleRawEvent);
      var _extra = extraRef.current;
      _extra && _forEachInstanceProperty(_context = _Object$entries(_extra)).call(_context, function (_ref7) {
        var _ref8 = _slicedToArray(_ref7, 2),
            key = _ref8[0],
            value = _ref8[1];

        if (key !== 'constructor' && key !== 'prototype' && key !== '__proto__') {
          recognition[key] = value;
        }
      });
      recognition.start();
    }

    return function () {
      var recognition = recognitionRef.current;

      if (recognition) {
        if (recognitionAbortable(recognition)) {
          recognition.abort();
        } else {
          throw new Error('Failed to stop recognition while the current one is ongoing and is not abortable.');
        }
      }
    };
  }, [extraRef, grammarRef, handleAudioEnd, handleAudioStart, handleEnd, handleError, handleRawEvent, handleResult, handleStart, langRef, notAllowedRef, recognitionRef, speechGrammarListRef, speechRecognitionRef, started]);
  var abortable = recognitionAbortable(recognitionRef.current) && readyState === 2;
  var supported = !!speechRecognition && !notAllowedRef.current;
  var context = useMemo(function () {
    return {
      abortable: abortable,
      readyState: readyState,
      supported: supported
    };
  }, [abortable, readyState, supported]);
  return /*#__PURE__*/React.createElement(Context.Provider, {
    value: context
  }, /*#__PURE__*/React.createElement(Context.Consumer, null, function (context) {
    return typeof children === 'function' ? children(context) : children;
  }));
};

Composer.defaultProps = {
  children: undefined,
  extra: undefined,
  grammar: undefined,
  lang: undefined,
  onDictate: undefined,
  onError: undefined,
  onProgress: undefined,
  onRawEvent: undefined,
  speechGrammarList: navigator.mediaDevices && navigator.mediaDevices.getUserMedia && vendorPrefix('SpeechGrammarList'),
  speechRecognition: navigator.mediaDevices && navigator.mediaDevices.getUserMedia && vendorPrefix('SpeechRecognition'),
  started: undefined
};
Composer.propTypes = {
  children: PropTypes.any,
  extra: PropTypes.any,
  grammar: PropTypes.string,
  lang: PropTypes.string,
  onDictate: PropTypes.func,
  onError: PropTypes.func,
  onProgress: PropTypes.func,
  onRawEvent: PropTypes.func,
  speechGrammarList: PropTypes.any,
  speechRecognition: PropTypes.any,
  started: PropTypes.any
};
export default Composer;
