// botframework-webchat/lib/adaptiveCards/AttachmentForScreenReader/RichCardAttachment.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _botframeworkWebchatComponent = require("botframework-webchat-component");

var _useUniqueId = _interopRequireDefault(require("../hooks/internal/useUniqueId"));

/* eslint-disable react/forbid-dom-props */

/* eslint-disable react/no-array-index-key */
var useLocalizer = _botframeworkWebchatComponent.hooks.useLocalizer;

var RichCardAttachment = function RichCardAttachment(_ref) {
  var _ref$content = _ref.content,
      content = _ref$content === void 0 ? {} : _ref$content;
  var labelId = (0, _useUniqueId.default)('webchat__id');
  var localize = useLocalizer();
  var buttons = content.buttons,
      facts = content.facts,
      image = content.image,
      images = content.images,
      items = content.items,
      subtitle = content.subtitle,
      tax = content.tax,
      text = content.text,
      title = content.title,
      total = content.total,
      vat = content.vat;
  var taxLabel = localize('RECEIPT_CARD_TAX');
  var totalLabel = localize('RECEIPT_CARD_TOTAL');
  var vatLabel = localize('RECEIPT_CARD_VAT');
  var cardLabel = localize('ATTACHMENT_CARD', title || '', subtitle || '', text || '');
  return /*#__PURE__*/_react.default.createElement("article", {
    "aria-labelledby": labelId
  }, /*#__PURE__*/_react.default.createElement("div", {
    id: labelId
  }, cardLabel), !!image && !!image.alt && /*#__PURE__*/_react.default.createElement("img", {
    alt: image.alt
  }), !!images && !!images.length && images.map(function (_ref2, index) {
    var alt = _ref2.alt;
    return /*#__PURE__*/_react.default.createElement("img", {
      alt: alt,
      key: index
    });
  }), !!facts && !!facts.length && /*#__PURE__*/_react.default.createElement("dl", null, facts.map(function (_ref3, index) {
    var key = _ref3.key,
        value = _ref3.value;
    return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, {
      key: index
    }, /*#__PURE__*/_react.default.createElement("dt", null, key), /*#__PURE__*/_react.default.createElement("dd", null, value));
  })), !!items && !!items.length && /*#__PURE__*/_react.default.createElement("ul", null, items.map(function (_ref4, index) {
    var image = _ref4.image,
        price = _ref4.price,
        quantity = _ref4.quantity,
        subtitle = _ref4.subtitle,
        text = _ref4.text,
        title = _ref4.title;
    return /*#__PURE__*/_react.default.createElement("li", {
      key: index
    }, !!image && !!image.alt && /*#__PURE__*/_react.default.createElement("img", {
      alt: image.alt
    }), !!title && /*#__PURE__*/_react.default.createElement("p", null, title), !!subtitle && /*#__PURE__*/_react.default.createElement("p", null, subtitle), !!text && /*#__PURE__*/_react.default.createElement("p", null, text), !!quantity && /*#__PURE__*/_react.default.createElement("p", null, quantity), !!price && /*#__PURE__*/_react.default.createElement("p", null, price));
  })), !!vat && /*#__PURE__*/_react.default.createElement("p", null, vatLabel, " ", vat), !!tax && /*#__PURE__*/_react.default.createElement("p", null, taxLabel, " ", tax), !!total && /*#__PURE__*/_react.default.createElement("p", null, totalLabel, " ", total), !!buttons && !!buttons.length && /*#__PURE__*/_react.default.createElement("div", null, buttons.map(function (_ref5, index) {
    var title = _ref5.title;
    return /*#__PURE__*/_react.default.createElement("button", {
      key: index,
      tabIndex: -1,
      type: "button"
    }, title);
  })));
};

RichCardAttachment.propTypes = {
  content: _propTypes.default.any.isRequired
};
var _default = RichCardAttachment;
exports.default = _default;
