// botframework-webchat-component/lib/Styles/StyleSet/SendBox.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createSendBoxStyle;

function createSendBoxStyle(_ref) {
  var sendBoxBackground = _ref.sendBoxBackground,
      sendBoxBorderBottom = _ref.sendBoxBorderBottom,
      sendBoxBorderLeft = _ref.sendBoxBorderLeft,
      sendBoxBorderRight = _ref.sendBoxBorderRight,
      sendBoxBorderTop = _ref.sendBoxBorderTop,
      sendBoxHeight = _ref.sendBoxHeight;
  return {
    '&.webchat__send-box': {
      '& .webchat__send-box__button--align-bottom': {
        alignSelf: 'flex-end'
      },
      '& .webchat__send-box__button--align-stretch': {
        alignSelf: 'stretch'
      },
      '& .webchat__send-box__button--align-top': {
        alignSelf: 'flex-start'
      },
      '& .webchat__send-box__main': {
        alignItems: 'stretch',
        backgroundColor: sendBoxBackground,
        borderBottom: sendBoxBorderBottom,
        borderLeft: sendBoxBorderLeft,
        borderRight: sendBoxBorderRight,
        borderTop: sendBoxBorderTop,
        minHeight: sendBoxHeight
      }
    }
  };
}
